# DATA CLEANING ON FIFA_21 DATASET



###**Project Overview**

**Introduction:**
This project aims to clean and preprocess the FIFA player dataset, which contains detailed attributes and skills of soccer players.


**Project Background:**
The FIFA dataset contains information on various aspects of soccer players, teams, and information of their matches such as team statistics, and match results.


**Objective:**
The sole objective of this project is to perform data cleaning: handling missing values, removing duplicates, handling inconsistent data, and transforming the data into a suitable format for analysis, visualization and modeling tasks.


**Dataset Description:**
The FIFA dataset consists of 18979 rows and 77 columns, including:

- Player information (ID, Name, LongName, photoUrl, playerUrl, Nationality, Age)
- Performance attributes (OVA, POT, Club, Contract, Positions, Height, Weight, Preferred Foot)
- Skills and abilities (Attacking, Crossing, Finishing, Heading Accuracy, etc.)
- Goalkeeping attributes (GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes)
- Physical attributes (Acceleration, Sprint Speed, Agility, Reactions, Balance, Power)
- Mental attributes (Mentality, Aggression, Interceptions, Positioning, Vision, Penalties)
- Defensive attributes (Defending, Marking, Standing Tackle, Sliding Tackle)
- Total Stats, Base Stats, and other miscellaneous attributes (W/F, SM, A/W, D/W, IR, PAC, SHO, PAS, DRI, DEF, PHY, Hits)


**Goals of Data Cleaning Phase:**

- Handle missing values
- Remove duplicates and redundant data
- Correct errors and inconsistencies
- Transform data into a suitable format for analysis
- Ensure data quality and integrity

By achieving these goals, we will produce a refined and reliable dataset ready for analysis, visualization, and modeling tasks.


## **Import the necessay libraries**

In [2]:
import pandas as pd
import numpy as np
import re # for regex expression
import datetime as dt

## **Import Dataset**

In [4]:
df = pd.read_csv("fifa21 raw data v2.csv")

# first 5 rows
df.head()

<ipython-input-4-65aa585334af>:1: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("fifa21 raw data v2.csv")


ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

In [5]:
# last 5 rows
df.tail()

ID          Name        LongName  \
18974  247223        Xia Ao          Ao Xia   
18975  258760      B. Hough       Ben Hough   
18976  252757   R. McKinley  Ronan McKinley   
18977  243790  Wang Zhen'ao    Zhen'ao Wang   
18978  252520     Zhou Xiao       Xiao Zhou   

                                               photoUrl  \
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                      Club  ...     A/W     D/W   IR PAC SHO  \
18974    47   55        \n\n\n\nWuhan Zall  ...  Medium  Medium  1 ★  64  28   
18975    47   67   \n\n\n\nOldham Athletic  ...  Medium  Medium  1 ★  64  40   
18976    47   65        \n\n\n\nDerry City  ...  Medium  Medium  1 ★  63  39   
18977    47   57  \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  58  49   
18978    47   57  \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  62  22   

       PAS DRI DEF PHY Hits  
18974   26  38  48  51  NaN  
18975   48  49  35  45  NaN  
18976   44  46  40  53  NaN  
18977   41  49  30  44  NaN  
18978   39  42  45  55  NaN  

[5 rows x 77 columns]

In [6]:
# Dimension of the dataset
df.shape

(18979, 77)

In [7]:
# Information of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

The dataset's information reveals a mix of data types, comprising both object and integer values, and also indicates the presence of null and non-null features, providing a comprehensive overview of the data's structure and content.

In [8]:
# Summary Statistics of the dataset
df.describe().T

count           mean           std    min       25%  \
ID                18979.0  226403.384794  27141.054157   41.0  210135.0   
Age               18979.0      25.194109      4.710520   16.0      21.0   
↓OVA              18979.0      65.718636      6.968999   47.0      61.0   
POT               18979.0      71.136414      6.114635   47.0      67.0   
BOV               18979.0      66.751726      6.747193   48.0      62.0   
Attacking         18979.0     248.938142     74.299428   42.0     222.0   
Crossing          18979.0      49.688392     18.131153    6.0      38.0   
Finishing         18979.0      45.842405     19.567081    3.0      30.0   
Heading Accuracy  18979.0      51.942726     17.294409    5.0      44.0   
Short Passing     18979.0      58.768112     14.519106    7.0      54.0   
Volleys           18979.0      42.696507     17.646937    3.0      30.0   
Skill             18979.0     256.479214     78.650601   40.0     222.0   
Dribbling         18979.0      55.587491     18.761314    5.0      49.0   
Curve             18979.0      47.246957     18.207790    4.0      35.0   
FK Accuracy       18979.0      42.391011     17.227947    5.0      31.0   
Long Passing      18979.0      52.725381     15.178151    5.0      43.0   
Ball Control      18979.0      58.528373     16.565892    5.0      54.0   
Movement          18979.0     317.718689     55.879046  122.0     289.0   
Acceleration      18979.0      64.357553     14.889981   13.0      57.0   
Sprint Speed      18979.0      64.412930     14.638743   12.0      57.0   
Agility           18979.0      63.366721     14.596277   14.0      55.0   
Reactions         18979.0      61.655619      9.072114   24.0      56.0   
Balance           18979.0      63.925865     14.062285   12.0      56.0   
Power             18979.0     296.605195     50.732390  122.0     264.0   
Shot Power        18979.0      57.801676     13.308747   18.0      48.0   
Jumping           18979.0      64.577375     11.844870   15.0      58.0   
Stamina           18979.0      62.653986     15.804223   12.0      55.0   
Strength          18979.0      64.786922     12.488672   16.0      57.0   
Long Shots        18979.0      46.785236     19.300534    4.0      32.0   
Mentality         18979.0     254.199747     64.595613   50.0     227.5   
Aggression        18979.0      55.591285     17.140414    9.0      44.0   
Interceptions     18979.0      46.394489     20.698078    3.0      25.0   
Positioning       18979.0      50.298962     19.428701    2.0      40.0   
Vision            18979.0      53.857105     13.708574    9.0      45.0   
Penalties         18979.0      48.057906     15.655999    6.0      39.0   
Composure         18979.0      58.073660     12.079110   12.0      50.0   
Defending         18979.0     139.909268     61.219472   20.0      83.0   
Marking           18979.0      46.559197     20.139324    3.0      29.0   
Standing Tackle   18979.0      47.705622     21.367690    5.0      27.0   
Sliding Tackle    18979.0      45.644449     20.922087    4.0      25.0   
Goalkeeping       18979.0      81.436693     84.665579   10.0      48.0   
GK Diving         18979.0      16.416302     17.554194    2.0       8.0   
GK Handling       18979.0      16.207124     16.816305    2.0       8.0   
GK Kicking        18979.0      16.076453     16.491103    2.0       8.0   
GK Positioning    18979.0      16.217187     17.002239    2.0       8.0   
GK Reflexes       18979.0      16.519627     17.854079    2.0       8.0   
Total Stats       18979.0    1595.286949    269.874789  747.0    1452.0   
Base Stats        18979.0     355.702197     40.761117  232.0     327.0   
PAC               18979.0      67.453975     10.677859   25.0      61.0   
SHO               18979.0      53.457031     13.827425   16.0      44.0   
PAS               18979.0      57.681016     10.081857   25.0      51.0   
DRI               18979.0      62.875020      9.927415   25.0      57.0   
DEF               18979.0      49.866221     

The summary statistics above displays the count, mean, standard deviation, minimum, percentile and maximum values of all numeric data.

In [9]:
# Check the column names
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

In [10]:
# Drop Unnecessary redundant columns
df.drop(columns = ["Name", "photoUrl", "playerUrl"], inplace = True)

We removed the columns because they lacked relevant information and/or are redundant in the dataset.

In [11]:
# Renaming the columns to improve readability and understanding of the data
df = df.rename(columns= {"LongName":"Name","↓OVA":"Overall_Rating",
                "POT": "Potential_Rating", "W/F": "Weak_Foot",
               "SM": "Skill_Moves", "A/W":"Attack_Willingness", "FK Accuracy" : "Free_Kick_Accuracy",
                "GK Diving" :"Goal_Keeper_Diving",
               "PAC": "Passing", "DRI":"Dribbling", "DEF": "Defending",
               "PHY": "Physicality", "BOV": "Ball_Outreach_Vision"})

## **Handling Missing Values**

In [12]:
# Number of missing values
df.isnull().sum().to_dict()

# Alternatively, df.isnull().to_frame().transpose()

{'ID': 0,
 'Name': 0,
 'Nationality': 0,
 'Age': 0,
 'Overall_Rating': 0,
 'Potential_Rating': 0,
 'Club': 0,
 'Contract': 0,
 'Positions': 0,
 'Height': 0,
 'Weight': 0,
 'Preferred Foot': 0,
 'Ball_Outreach_Vision': 0,
 'Best Position': 0,
 'Joined': 0,
 'Loan Date End': 17966,
 'Value': 0,
 'Wage': 0,
 'Release Clause': 0,
 'Attacking': 0,
 'Crossing': 0,
 'Finishing': 0,
 'Heading Accuracy': 0,
 'Short Passing': 0,
 'Volleys': 0,
 'Skill': 0,
 'Dribbling': 0,
 'Curve': 0,
 'Free_Kick_Accuracy': 0,
 'Long Passing': 0,
 'Ball Control': 0,
 'Movement': 0,
 'Acceleration': 0,
 'Sprint Speed': 0,
 'Agility': 0,
 'Reactions': 0,
 'Balance': 0,
 'Power': 0,
 'Shot Power': 0,
 'Jumping': 0,
 'Stamina': 0,
 'Strength': 0,
 'Long Shots': 0,
 'Mentality': 0,
 'Aggression': 0,
 'Interceptions': 0,
 'Positioning': 0,
 'Vision': 0,
 'Penalties': 0,
 'Composure': 0,
 'Defending': 0,
 'Marking': 0,
 'Standing Tackle': 0,
 'Sliding Tackle': 0,
 'Goalkeeping': 0,
 'Goal_Keeper_Diving': 0,
 'GK Handl

With the exception of "Loan Date End" and "Hits", which have 17,966 and 2,595 missing values respectively, all features have complete data. To address this, we can either drop these columns or replace the missing values with suitable alternatives such as mean, mode, median, or other relevant values. For the purpose of data cleaning, we will opt to replace the missing values to maintain data integrity and facilitate analysis.

### Handling Missing Data in the Loan Date End Feature

In [13]:
# Count the missing and non missing values
df["Loan Date End"].isna().value_counts()

Loan Date End
True     17966
False     1013
Name: count, dtype: int64

In [14]:
# Unique values
df["Loan Date End"].unique()

array([nan, 'Jun 30, 2021', 'Dec 31, 2020', 'Jan 30, 2021',
       'Jun 30, 2022', 'May 31, 2021', 'Jul 5, 2021', 'Dec 31, 2021',
       'Jul 1, 2021', 'Jan 1, 2021', 'Aug 31, 2021', 'Jan 31, 2021',
       'Dec 30, 2021', 'Jun 23, 2021', 'Jan 3, 2021', 'Nov 27, 2021',
       'Jan 17, 2021', 'Jun 30, 2023', 'Jul 31, 2021', 'Nov 22, 2020',
       'May 31, 2022', 'Dec 30, 2020', 'Jan 4, 2021', 'Nov 30, 2020',
       'Aug 1, 2021'], dtype=object)

We are to conduct a cross-validation exercise to verify the missing values in the "Loan Date End" feature by cross-checking them against the "Contract" and "Joined" features, seeking any potential correlations or matches, and printing the results to ensure data accuracy and consistency.

In [15]:
# filter out the nan rows in loan date end column
filtered_null_loan = df[df["Loan Date End"].isna()]

# Print
print(f"\n The null values:")
print(filtered_null_loan[["Loan Date End","Contract","Joined" ]].sample(10))


 The null values:
      Loan Date End     Contract        Joined
16399           NaN  2020 ~ 2020  Jul 31, 2020
18584           NaN  2020 ~ 2024   Jan 1, 2020
6542            NaN  2019 ~ 2023   Jan 1, 2019
14627           NaN  2019 ~ 2022   Mar 9, 2019
18865           NaN  2019 ~ 2024   Jan 1, 2019
10473           NaN  2020 ~ 2020   Jan 7, 2020
4030            NaN         Free   Feb 6, 2020
13808           NaN  2020 ~ 2023  Jul 15, 2020
15541           NaN  2017 ~ 2021  Jan 31, 2017
14971           NaN  2019 ~ 2024  Jun 24, 2019


In [19]:
# crosscheck the contract and loan date end columns for correspondance
filtered_not_null =df[df["Loan Date End"].notna()]

# print result
print(f"\n The not-null values:""\n")
print(filtered_not_null[["Loan Date End", "Contract", "Joined"]].sample(20))


 The not-null values:

      Loan Date End              Contract        Joined
14837  Jun 30, 2022  Jun 30, 2022 On Loan   Jul 1, 2019
1281   Jun 30, 2021  Jun 30, 2021 On Loan  Aug 22, 2015
14759  Jun 30, 2021  Jun 30, 2021 On Loan   Aug 1, 2019
7275   Jun 30, 2021  Jun 30, 2021 On Loan   Jul 1, 2019
13108  Jun 30, 2021  Jun 30, 2021 On Loan   Jul 5, 2019
17770  Nov 30, 2020  Nov 30, 2020 On Loan  Mar 12, 2018
446    Jun 30, 2021  Jun 30, 2021 On Loan  Jun 11, 2018
770    Jun 30, 2021  Jun 30, 2021 On Loan  Aug 31, 2017
7284   Dec 31, 2020  Dec 31, 2020 On Loan   Aug 7, 2018
7225   Jun 30, 2021  Jun 30, 2021 On Loan  Jan 25, 2019
3379   Jun 30, 2021  Jun 30, 2021 On Loan   Jul 1, 2016
13910  Jun 30, 2021  Jun 30, 2021 On Loan   Jul 2, 2018
3086   Jun 30, 2021  Jun 30, 2021 On Loan  Jul 16, 2018
13102  Jun 30, 2021  Jun 30, 2021 On Loan  Aug 27, 2018
13089  Jun 30, 2021  Jun 30, 2021 On Loan  Jun 24, 2016
6768   Jun 30, 2021  Jun 30, 2021 On Loan   Aug 9, 2018
4276   Dec 31, 2020  Dec

Upon observing that the 'Contract' feature dates align with the 'Loan Date End' feature, we can conclude that the missing 'Loan Date End' values can be reliably replaced with the corresponding 'Contract' values, ensuring data consistency and accuracy. However, since the dates are identical and the 'Contract' feature also indicates loan status, we opt to drop the redundant 'Loan Date End' feature, preserving data efficiency and clarity.

In [20]:
# drop Loan Date End column
df = df.drop(columns=["Loan Date End"])

### Imputing Missing Values in the Hits Feature




In [21]:
# Count the missing and non missing values
df["Hits"].isna().value_counts()

Hits
False    16384
True      2595
Name: count, dtype: int64

In [22]:
# Unique Values
df["Hits"].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

The null values in the "Hits" feature will be replaced with zero, signifying no tackles or contact, thereby acknowledging potential non-participation in plays or lack of recorded tackles, and providing a more comprehensive understanding of player performance by accounting for these instances.

In [39]:
# Replace with Zero
df["Hits"] = df["Hits"].fillna(0)

# Checking if the changes had the desired impact
df["Hits"].isna().sum()


0

With the successful replacement of null values in the "Hits" and "Loan Date End" features, we have effectively addressed the issue of missing data, ensuring a more complete and accurate dataset.

## **Checking Duplicates**

In [24]:
# Check for duplicates
df.duplicated().any()

False

Since there are no duplicates, we move on to the next cleaning phase.

## **Standardizing Data Format**

"Next, we will standardize the data by transforming columns into a consistent and uniform format, making it more efficient to work with and analyze.
These features include:

*  Height and Weight
*  Hits
*  Club
*  Value, Wage, Release Clause
*  IR
*  Name
* Contract








### A) Convert Height and Weight To Numerical data

#### Height Feature:

In [26]:
# Info of the features
df[["Height", "Weight"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Height  18979 non-null  object
 1   Weight  18979 non-null  object
dtypes: object(2)
memory usage: 296.7+ KB


In [27]:
# Display unique values within feature
df["Height"].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

The "Height" feature, currently stored as an object data type, exhibits inconsistent formatting, with values expressed in both centimeters (cm) and feet & inches. To achieve numerical uniformity, we must first remove the unit labels (cm) and then convert the feet and inches values to centimeters, enabling a successful conversion to a numerical data type.

In [28]:
# Replace cm with " "
df["Height"] =  df["Height"].str.replace("cm", "")

In [29]:
# function to convert height in feet and inches to cm
def convert_to_cm(x):
  """
  Converts a height measurement to centimeters.

  Parameters:
  x (str or float): The height measurement to convert.

  Returns:
  float: The height measurement in centimeters .
  """
  if '\'' in x:
        feet, inches = x.split('\'')
        inches = inches.replace('"', '')
        total_inches = int(feet) * 12 + int(inches)
        cm = total_inches * 2.54
        return cm
  else:
        return float(x)

# Apply the function to the Height column
df['Height'] = df['Height'].apply(convert_to_cm)

# Convert the Height column to integer
df['Height'] = df['Height'].astype(int)

In [30]:
# Verifying the effectiveness of the changes
df['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 162,
       201, 158, 161, 160, 203, 157, 156, 202, 159, 206, 155])

#### Weight Feature:

In [31]:
df["Weight"].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

The "Weight" feature contains values expressed in both kilograms (kg) and pounds (lbs), necessitating a conversion to a standardized unit. Specifically, we need to convert the pounds values to kilograms, extract the numerical values by removing the unit labels, and finally convert the column to a numerical data type.




In [32]:
# Function to convert Ibs to kg
def convert_to_kg(column, kg = 0.453592):
   """
   Converts a weight value from pounds (lbs) to kilograms (kg).

   Parameters:
   column (str or numeric): The weight value to convert. Can be a string containing "lbs" or a numeric value.
   kg (float, optional): The conversion factor from pounds to kilograms. Default is 0.453592.

   Returns:
   float: The converted weight value in kilograms.

   Notes:
   If the input is a string that doesn't contain "lbs", it is returned unchanged.
   """
   if isinstance(column, str):
        if "lbs" in column:
            con_kg = float(column.replace("lbs", "")) * kg
            return con_kg
        elif "kg" in column:
            return float(column.replace("kg", ""))
   return column

# Applying the function to the "Weight" column
df["Weight"] = df["Weight"].apply(convert_to_kg)

# Converting to int
df["Weight"] =  df["Weight"].astype(int)


In [33]:
# Verifying the effectiveness of the changes
df["Weight"].unique()

array([ 72,  83,  87,  70,  68,  80,  71,  91,  73,  85,  92,  69,  84,
        96,  81,  82,  75,  86,  89,  74,  76,  64,  78,  90,  66,  60,
        94,  79,  67,  65,  59,  61,  93,  88,  97,  77,  62,  63,  95,
       100,  58,  98, 103,  99, 102,  56, 101,  57,  55, 104, 107, 110,
        53,  50,  54,  52])

In [34]:
# Modify the columns name to show the unit of measurement at the header
df = df.rename(columns ={"Height": "Height(cm)",
                    "Weight": "Weight(kg)"})

#### Hits Feature:

In [35]:
df["Hits"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 18979 entries, 0 to 18978
Series name: Hits
Non-Null Count  Dtype 
--------------  ----- 
18979 non-null  object
dtypes: object(1)
memory usage: 148.4+ KB


In [36]:
df["Hits"].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

The "Hits" feature, currently represented as an object data type, requires standardization and cleaning. Some values in this column are expressed in thousands, denoted by the abbreviation "k", and also float values. To ensure consistency, we need to remove the "k" notation, multiply the values by 1000, and convert the column to an integer data type. This will enable us to uniformly represent the number of tackles made by a player in a season, allowing for accurate analysis and processing.





In [41]:
def convert_to_thousand(x):
  """
    Converts a string value to a float,
    multiplying by 1000 if the value is in thousands (denoted by "K").

    Parameters:

    - x (string): The value to be converted.

    Returns:

    - thou_hits (float): The converted value.
  """
  if type(x) == str:  # Check if x is a string
        if "K" in x:
            thou_hits = float(x.replace("K"," ")) * 1000
            return thou_hits
  else:
    return float(x)

# Applying the function to the "Hits" column
df["Hits"] = df["Hits"].apply(convert_to_thousand)

In [42]:
# Convert data type to int
df["Hits"] = df["Hits"].astype(int)

# Verifying the effectiveness of the changes
df["Hits"].unique()

array([   0, 1600, 1100, 1500, 3200, 1700, 2300, 1900, 2000, 1300, 3000,
       1400, 2800, 1800, 8400, 4300, 1200, 1000, 2500, 6000,   11,    2,
          1,   31,    3,   10,    9,   17,    7,    4,    6])

#### Club Feature:

In [43]:
# Unique values in Club features
df["Club"].unique()

array(['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus',
       '\n\n\n\nAtlético Madrid', '\n\n\n\nManchester City',
       '\n\n\n\nParis Saint-Germain', '\n\n\n\nFC Bayern München',
       '\n\n\n\nLiverpool', '\n\n\n\nReal Madrid', '\n\n\n\nChelsea',
       '\n\n\n\nTottenham Hotspur', '\n\n\n\nInter', '\n\n\n\nNapoli',
       '\n\n\n\nBorussia Dortmund', '\n\n\n\nManchester United',
       '\n\n\n\nArsenal', '\n\n\n\nLazio', '\n\n\n\nLeicester City',
       '\n\n\n\nBorussia Mönchengladbach', '\n\n\n\nReal Sociedad',
       '\n\n\n\nAtalanta', '\n\n\n\nOlympique Lyonnais', '\n\n\n\nMilan',
       '\n\n\n\nVillarreal CF', '\n\n\n\nRB Leipzig', '\n\n\n\nCagliari',
       '\n\n\n\nAjax', '\n\n\n\nSL Benfica', '\n\n\n\nAS Monaco',
       '\n\n\n\nWolverhampton Wanderers', '\n\n\n\nEverton',
       '\n\n\n\nFiorentina', '\n\n\n\nFC Porto', '\n\n\n\nRC Celta',
       '\n\n\n\nTorino', '\n\n\n\nSevilla FC', '\n\n\n\nGrêmio',
       '\n\n\n\nReal Betis', '\n\n\n\nRoma', '\n\n\n\nNewcastle Unite

In [44]:
# A random club feature infomation
df["Club"].sample(10)

4926                               \n\n\n\nFulham
14271                              \n\n\n\nBarrow
14943                         \n\n\n\nVissel Kobe
13583                       \n\n\n\nHallescher FC
4913                           \n\n\n\nCD Leganés
10440                    \n\n\n\nKFC Uerdingen 05
12674    \n\n\n\nUniversidad Católica del Ecuador
15096                          \n\n\n\nGwangJu FC
19                                \n\n\n\nChelsea
5394                        \n\n\n\nMiddlesbrough
Name: Club, dtype: object

It is apparent that the players are affiliated with various clubs, and the associated column requires cleaning. To achieve this, we need to perform the following tasks:

- Remove newline characters (\n) to consolidate the data into a single line
- Trim trailing spaces to ensure consistent formatting



In [45]:
# Remove the newlines
df["Club"] = df["Club"].str.replace("\n", " ")

# Determining if the changes were successful
print(f'\nThe unique clubs are:\n')
df["Club"].unique()


The unique clubs are:



array(['    FC Barcelona', '    Juventus', '    Atlético Madrid',
       '    Manchester City', '    Paris Saint-Germain',
       '    FC Bayern München', '    Liverpool', '    Real Madrid',
       '    Chelsea', '    Tottenham Hotspur', '    Inter', '    Napoli',
       '    Borussia Dortmund', '    Manchester United', '    Arsenal',
       '    Lazio', '    Leicester City', '    Borussia Mönchengladbach',
       '    Real Sociedad', '    Atalanta', '    Olympique Lyonnais',
       '    Milan', '    Villarreal CF', '    RB Leipzig', '    Cagliari',
       '    Ajax', '    SL Benfica', '    AS Monaco',
       '    Wolverhampton Wanderers', '    Everton', '    Fiorentina',
       '    FC Porto', '    RC Celta', '    Torino', '    Sevilla FC',
       '    Grêmio', '    Real Betis', '    Roma', '    Newcastle United',
       '    Eintracht Frankfurt', '    Valencia CF',
       '    Medipol Başakşehir FK', '    Inter Miami',
       '    Bayer 04 Leverkusen', '    Levante UD', '    Crystal 

In [55]:
# Replace digits attached to club names with empty string
df["Club"] = df["Club"].str.replace(r'\d+', '', regex=True)

# Replace dots attached to club names with empty string
df["Club"] = df["Club"].str.replace(r'. ', '', regex=True)

# Remove the leading and trailing characters (spaces, tabs, etc.)
df["Club"] = df["Club"].str.strip()

# Assessing the impact of the changes
df["Club"].sample(5)

4983          Guadalajara
11698    Waasland-Beveren
16788            Sassuolo
5037               MalmFF
314              SBenfica
Name: Club, dtype: object

#### Value, Wage and Release Clause Features:


In [56]:
# Check datatype
data_type = df[["Value", "Wage","Release Clause"]].dtypes

# Print datatype
print(f"\nThe datatypes are:")

print(data_type)

# Assessing values within the features
random_rows = df[["Value", "Wage","Release Clause"]].sample(10)

# Print
print(f"\n{random_rows}")


The datatypes are:
Value             object
Wage              object
Release Clause    object
dtype: object

       Value  Wage Release Clause
13098  €975K   €6K             €0
7035   €1.4M   €5K          €1.5M
7779   €1.7M   €5K          €1.8M
10626  €825K   €3K          €953K
7949   €1.3M   €8K          €1.2M
18551  €210K   €5K          €278K
10048  €675K   €3K          €1.1M
6905     €1M  €550          €1.6M
13640  €875K   €2K          €878K
18786  €150K  €500          €221K


These features contain monetary values, currently in object dtype, and require conversion to numerical dtype for analysis. To achieve this, we need to:

*  Remove the euro symbol (€) from the values
*  Remove the thousand separator ('k') and multiply the values by 1000
   to convert to the correct scale.



In [57]:
# Remove the euro symbol (€)
df["Value"] = df["Value"].map(lambda x: x.replace("€", ""))

df["Wage"] = df["Wage"].map(lambda x: x.replace("€", ""))

df["Release Clause"] = df["Release Clause"].map(lambda x: x.replace("€", ""))

# Assessing the update
df[["Value", "Wage","Release Clause"]]


Value  Wage Release Clause
0      103.5M  560K         138.4M
1         63M  220K          75.9M
2        120M  125K         159.4M
3        129M  370K           161M
4        132M  270K         166.5M
...       ...   ...            ...
18974    100K    1K            70K
18975    130K   500           165K
18976    120K   500           131K
18977    100K    2K            88K
18978    100K    1K            79K

[18979 rows x 3 columns]

In [58]:
 # Function to convert monetary values to numerical
def convert_to_numeric(value, t = 1000, m = 1000000, b = 1000000000):
  """
    Converts a string a numeric value, handling thousand (K), million (M), and billion (B) suffixes.

    Parameters:
    value (str): The value to convert.
    t (int, optional): The multiplier for thousand (default is 1000).
    m (int, optional): The multiplier for million (default is 1000000).
    b (int, optional): The multiplier for billion (default is 1000000000).

    Returns:
    float: The converted numeric value.

   Notes:
   If the input is a string, it may contain a suffix (K, M, or B) indicating the scale.
   If the input is already a numeric value, it is returned unchanged.
  """

  if type(value) == str:
     if "K" in value:
        return float(value.replace("K", " "))* t
     elif "M" in value:
        return float(value.replace("M", " "))* m
     elif "B" in value:
        return float(value.replace("B", " "))* b
     else:
        return float(value)
  return value



# columns to convert
columns_to_convert = ["Value", "Wage", "Release Clause"]

# Loop through the columns
for column in columns_to_convert:

# Apply the function to the columns
  df[column] = df[column].apply(convert_to_numeric)


In [59]:
# Verifying the effectiveness of the updates
df[["Value", "Wage", "Release Clause"]].dtypes

Value             float64
Wage              float64
Release Clause    float64
dtype: object

In [60]:
# Properly insert the euro symbol (€) into the column header
df = df.rename(columns={"Value": "Value(€)", "Wage": "Wage(€)", "Release Clause": "Release Clause(€)"})

# Assessing the impact of the changes
df[["Value(€)", "Wage(€)", "Release Clause(€)"]].sample(5)

Value(€)  Wage(€)  Release Clause(€)
441   24000000.0  31000.0         35900000.0
4207    650000.0   6000.0          1300000.0
6226   1700000.0   6000.0          1900000.0
5234   1300000.0   8000.0          2900000.0
9771   1400000.0   2000.0          1100000.0

#### IR  Feature:
This column actually represents player ratings or rankings, with the number of stars (★) indicating the level of performance.



In [61]:
df["IR"].unique()

array(['5 ★', '3 ★', '4 ★', '2 ★', '1 ★'], dtype=object)

The column contains player ratings in the format of '★', and we need to remove the ★ symbol from the values, leaving only the numerical rating.

In [62]:
# Remove ★
df["IR"] = df["IR"].str.replace("★", " ").astype(int)

# Display 10 random values
df["IR"].sample(10)

5220     1
8015     1
12933    1
898      2
481      2
2401     1
5256     1
15254    1
1895     1
6415     1
Name: IR, dtype: int64

#### Name Feature:

In [63]:
# First 5 names
df["Name"].head()

0                    Lionel Messi
1    C. Ronaldo dos Santos Aveiro
2                       Jan Oblak
3                 Kevin De Bruyne
4      Neymar da Silva Santos Jr.
Name: Name, dtype: object

In [64]:
# Capitalizes the first letter of each name
df["Name"] = df["Name"].str.title()

# Verying the update
df["Name"].head()

0                    Lionel Messi
1    C. Ronaldo Dos Santos Aveiro
2                       Jan Oblak
3                 Kevin De Bruyne
4      Neymar Da Silva Santos Jr.
Name: Name, dtype: object

#### Contract Feature:

In [65]:
# Unique values in Contract feature
df["Contract"].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

The column, intended to display only the end years of contracts, contains additional information, with some entries denoted as "on loan" for players on loan and "free" for players on a free transfer, as well as different date formats, requiring data cleaning and preprocessing to extract the relevant contract years.


In [66]:
# Retrive rows in the contract column that contains "On Loan"
loan_rows = df[df["Contract"].str.contains("On Loan")]
loan_rows

ID                       Name Nationality  Age  Overall_Rating  \
205    173731                Gareth Bale       Wales   30              83   
248    193105            Alphonse Areola      France   27              82   
254    200888  Danilo Luís Hélio Pereira    Portugal   28              82   
302    216409            Matteo Politano       Italy   26              81   
306    223959             Lucas Torreira     Uruguay   24              81   
...       ...                        ...         ...  ...             ...   
18472  242601               Moudi Najjar   Australia   20              52   
18571  236818                Arthur Read     England   20              51   
18600  240392                Yubiao Deng    China PR   23              51   
18622  252516                    Pu Chen    China PR   23              51   
18680  232888               Zhechao Chen    China PR   25              51   

       Potential_Rating                     Club              Contract  \
205                  83          TottenhaHotspur  Jun 30, 2021 On Loan   
248                  86                   Fulham  Jun 30, 2021 On Loan   
254                  82        PariSaint-Germain  Jun 30, 2021 On Loan   
302                  81                   Napoli  Jun 30, 2021 On Loan   
306                  85            AtléticMadrid  Jun 30, 2021 On Loan   
...                 ...                      ...                   ...   
18472                65               MacarthuFC  Aug 31, 2021 On Loan   
18571                63                Stevenage  Jun 30, 2021 On Loan   
18600                60  ShijiazhuanEveBrighF.C.  Dec 31, 2020 On Loan   
18622                60  ShijiazhuanEveBrighF.C.  Dec 31, 2020 On Loan   
18680                56             GuangzhoR&FC  Dec 31, 2020 On Loan   

        Positions  Height(cm)  ...  Attack_Willingness     D/W  IR Passing  \
205            RW         185  ...              Medium  Medium   4      88   
248            GK         195  ...              Medium  Medium   2      85   
254       CDM, CM         188  ...              Medium    High   3      66   
302    RM, RW, ST         171  ...                High  Medium   2      83   
306       CDM, CM         166  ...              Medium    High   2      74   
...           ...         ...  ...                 ...     ...  ..     ...   
18472          ST         182  ...                High     Low   1      67   
18571  CM, RW, LW         178  ...              Medium  Medium   1      57   
18600          CM         181  ...              Medium  Medium   1      71   
18622     CAM, LM         177  ...              Medium  Medium   1      59   
18680  CM, LM, LB         177  ...              Medium  Medium   1      61   

      SHO  PAS  Dribbling  Defending  Physicality  Hits  
205    84   82         80         58           73     0  
248    79   80         84         55           78     0  
254    63   72         73         80           85     0  
302    77   78         84         42           55     0  
306    67   75         79         79           74     0  
...    ..  ...        ...        ...          ...   ...  
18472  51   39         53         21           52     0  
18571  45   52         59         38           48     0  
18600  42   52         54         47           52     0  
18622  41   55         49         34           47     0  
18680  35   49         50         59           58     0  

[1013 rows x 73 columns]

1013 players are on loan

In [67]:
#Retrive rows in the contract column that contains "Free"
free_rows = df[df["Contract"].str.contains("Free")]
free_rows

ID                              Name           Nationality  Age  \
289    230347    Welington Kauê Dano Nascimento                Brazil   20   
292    230225  Juan Everton Mestres De Mesquita                Brazil   24   
369    245299            Jaime Nicolás Frendado               Uruguay   36   
374    245294          Jorge Ezequiel Serendero               Uruguay   32   
375    245308                Mauro Evidio Nérez               Uruguay   32   
...       ...                               ...                   ...  ...   
17220  247059                    Ali Al Hidhani  United Arab Emirates   22   
17343  251734                    Anudaan Sanghu                 India   20   
17659  251738                   Madanapal Chada                 India   27   
17661  251741                Eelamynthan Suresh                 India   28   
18205  251737                  Jagatjiva Mangat                 India   26   

       Overall_Rating  Potential_Rating   Club Contract    Positions  \
289                81                81  NClub     Free       LB, LM   
292                81                81  NClub     Free      CB, CDM   
369                80                80  NClub     Free      CB, CDM   
374                80                80  NClub     Free           GK   
375                80                80  NClub     Free       LB, LM   
...               ...               ...    ...      ...          ...   
17220              56                64  NClub     Free           RB   
17343              56                56  NClub     Free       RM, CM   
17659              55                55  NClub     Free           ST   
17661              55                55  NClub     Free  RM, CAM, RB   
18205              53                54  NClub     Free           LM   

       Height(cm)  ...  Attack_Willingness     D/W  IR Passing SHO  PAS  \
289           178  ...              Medium  Medium   1      79  54   76   
292           181  ...                High  Medium   1      71  61   57   
369           181  ...                High  Medium   1      71  61   57   
374           190  ...              Medium  Medium   1      78  81   77   
375           178  ...              Medium  Medium   1      79  54   75   
...           ...  ...                 ...     ...  ..     ...  ..  ...   
17220         172  ...              Medium  Medium   1      73  33   45   
17343         168  ...              Medium     Low   1      79  42   50   
17659         178  ...                High  Medium   1      67  52   47   
17661         175  ...                High  Medium   1      73  49   51   
18205         171  ...              Medium  Medium   1      68  41   45   

       Dribbling  Defending  Physicality  Hits  
289           78         75           74     0  
292           62         82           80     0  
369           62         82           75     0  
374           77         42           80     0  
375           78         76           74     0  
...          ...        ...          ...   ...  
17220         55         51           55     0  
17343         59         29           49     0  
17659         61         32           57     0  
17661         54         41           70     0  
18205         58         22           41     0  

[237 rows x 73 columns]

237 players are on free transfer

In [ ]:
df[["Contract", "Joined"]].sample(5)

Contract        Joined
6368            2019 ~ 2024   Jul 1, 2019
5792            2009 ~ 2020   Jan 1, 2009
17002           2019 ~ 2021   Jul 1, 2019
1884   Jun 30, 2021 On Loan   Jan 2, 2019
10394           2019 ~ 2023  Feb 24, 2019

We have to:

* Create new Status columns which returns True for players "On Loan" and vice-versa
* We will do same for players on "Free" transfer.
* Remove "On Loan" and "Free" information from the Contract column.
* Modify the Contract column into Contarct End Date.

In [68]:
def player_status(x):
  """
    Determine if a player was on loan transfer.

    Parameters:
    x (str): The contract information.

    Returns:
    bool: True if the player was signed on loan, False otherwise.
"""
  if "On Loan" in x:
      return True
  else:
      return False

# Apply the function to the Contract column to create the Player_on_Loan column
df["Player_on_Loan"] = df["Contract"].apply(player_status)

In [ ]:
df[["Contract", "Player_on_Loan"]].sample(15)

Contract  Player_on_Loan
6687            2019 ~ 2024           False
3019            2019 ~ 2021           False
1752            2018 ~ 2023           False
10707           2016 ~ 2022           False
4924            2010 ~ 2021           False
1676            2018 ~ 2022           False
1292            2018 ~ 2021           False
2729            2017 ~ 2022           False
17088           2018 ~ 2023           False
16029  Jun 30, 2021 On Loan            True
14113           2020 ~ 2022           False
8470   Jun 30, 2021 On Loan            True
3845            2020 ~ 2021           False
4737            2014 ~ 2021           False
12154           2014 ~ 2020           False

In [69]:
def free_transfer(x):
  """
    Determine if a player was signed on a free transfer.

    Parameters:
    x (str): The contract information.

    Returns:
    bool: True if the player was signed on a free transfer, False otherwise.
"""
  if "Free" in x:
        return True
  else:
        return False

# Apply the function to the Contract column to create the Player_on_Loan column
df["Free_transfer"] = df["Contract"].apply(free_transfer)

In [ ]:
df[["Contract", "Free_transfer"]].sample(10)

Contract  Free_transfer
842    2019 ~ 2021          False
11616  2019 ~ 2024          False
511    2017 ~ 2022          False
9901   2015 ~ 2022          False
14952         Free           True
16950  2019 ~ 2022          False
11779  2019 ~ 2021          False
3904   2020 ~ 2023          False
16611  2019 ~ 2021          False
5903   2020 ~ 2025          False

In [ ]:
# Remove On Loan and free string from contract column
df["Contract"]  = df["Contract"].str.replace("On Loan","")

df["Contract"]  = df["Contract"].str.replace("Free","")

# Verifying the update
df["Contract"].unique()


array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 ', '2009 ~ 2021',
       '2019 ~ 2021', '2019 ~ 2026', '', '2012 ~ 2028', '201

The Contract feature was meant to show contract duration (end year), but has inconsistent formatting. However, some rows incorrectly include the day and month.

In [75]:
# Function to clean the Contract column
def clean_contract(date_str):
   """
   Cleans a date string by removing month, day, and unnecessary characters.

   Parameters:
   date_str (str): The input date string to be cleaned.

   Returns:
   str: The cleaned date string with only the year remaining
   """
    # Remove month and date
   date_str = re.sub(r'[A-Za-z]+\s+\d{1,2},\s+', '', date_str)
    # Keep only the year before the ~ symbol
   date_str = re.sub(r'\s*~\s*\d{4}', '', date_str)
   return date_str.strip()

# Apply the function to the date column
df["Contract_End_Date"] = df["Contract"].apply(clean_contract)

In [76]:
df['Contract_End_Date'].sample(10)

18526            2016
9308     2021 On Loan
11386            2020
15047    2021 On Loan
1359             2019
14549            2020
18027            2019
967              2011
7844             2018
5365             Free
Name: Contract_End_Date, dtype: object

In [77]:
# Drop Contract column
df = df.drop(columns=["Contract"])

In [ ]:
# first 10 rows
df.head(10)

ID                          Name Nationality  Age  Overall_Rating  \
0  158023                  Lionel Messi   Argentina   33              93   
1   20801  C. Ronaldo Dos Santos Aveiro    Portugal   35              92   
2  200389                     Jan Oblak    Slovenia   27              91   
3  192985               Kevin De Bruyne     Belgium   29              91   
4  190871    Neymar Da Silva Santos Jr.      Brazil   28              91   
5  188545            Robert Lewandowski      Poland   31              91   
6  209331                 Mohamed Salah       Egypt   28              90   
7  212831         Alisson Ramses Becker      Brazil   27              90   
8  231747                 Kylian Mbappé      France   21              90   
9  192448         Marc-André Ter Stegen     Germany   28              90   

   Potential_Rating               Club   Positions  Height(cm)  Weight(kg)  \
0                93         FBarcelona  RW, ST, CF         170          72   
1                92           Juventus      ST, LW         187          83   
2                93      AtléticMadrid          GK         188          87   
3                91      ManchesteCity     CAM, CM         181          70   
4                91  PariSaint-Germain     LW, CAM         175          68   
5                91      FBayerMünchen          ST         184          80   
6                90          Liverpool          RW         175          71   
7                91          Liverpool          GK         191          91   
8                95  PariSaint-Germain  ST, LW, RW         178          73   
9                93         FBarcelona          GK         187          85   

   ... Passing  SHO PAS Dribbling  Defending  Physicality  Hits  \
0  ...      85   92  91        95         38           65     0   
1  ...      89   93  81        89         35           77     0   
2  ...      87   92  78        90         52           90     0   
3  ...      76   86  93        88         64           78     0   
4  ...      91   85  86        94         36           59     0   
5  ...      78   91  78        85         43           82     0   
6  ...      93   86  81        90         45           75     0   
7  ...      86   88  85        89         51           91     0   
8  ...      96   86  78        91         39           76  1600   
9  ...      88   85  88        90         45           88     0   

   Player_on_Loan  Free_transfer  Contract_Year_End  
0           False          False               2004  
1           False          False               2018  
2           False          False               2014  
3           False          False               2015  
4           False          False               2017  
5           False          False               2014  
6           False          False               2017  
7           False          False               2018  
8           False          False               2018  
9           False          False               2014  

[10 rows x 75 columns]

In [ ]:
# first 10 last rows
df.tail(10)

ID            Name Nationality  Age  Overall_Rating  \
18969  257936        Yue Song    China PR   28              47   
18970  257689    Huapeng Wang    China PR   20              47   
18971  255958    Mateo Flores     Bolivia   19              47   
18972  257710  Mengxuan Zhang    China PR   21              47   
18973  258736   Vani Da Silva     England   17              47   
18974  247223          Ao Xia    China PR   21              47   
18975  258760       Ben Hough     England   17              47   
18976  252757  Ronan Mckinley     England   18              47   
18977  243790    Zhen'Ao Wang    China PR   20              47   
18978  252520       Xiao Zhou    China PR   21              47   

       Potential_Rating                       Club Positions  Height(cm)  \
18969                47                TianjiTEDFC        CM         185   
18970                53               GuangzhoR&FC        CB         181   
18971                63                 CluBolívar       CDM         174   
18972                52  ChongqinDangdaLifaFSWTeam        CB         177   
18973                67              OldhaAthletic        ST         171   
18974                55                   WuhaZall        CB         178   
18975                67              OldhaAthletic        CM         175   
18976                65                   DerrCity        CM         179   
18977                57               DaliaYiFanFC        RW         175   
18978                57               DaliaYiFanFC    CB, LB         188   

       Weight(kg)  ... Passing  SHO PAS Dribbling  Defending  Physicality  \
18969          79  ...      65   37  49        46         40           59   
18970          77  ...      59   22  25        30         47           55   
18971          68  ...      57   32  43        48         44           49   
18972          70  ...      58   23  26        27         50           48   
18973          58  ...      70   46  40        53         16           40   
18974          66  ...      64   28  26        38         48           51   
18975          65  ...      64   40  48        49         35           45   
18976          74  ...      63   39  44        46         40           53   
18977          69  ...      58   49  41        49         30           44   
18978          75  ...      62   22  39        42         45           55   

       Hits  Player_on_Loan  Free_transfer  Contract_Year_End  
18969     0           False          False               2020  
18970     0           False          False               2020  
18971     0           False          False               2020  
18972     0           False          False               2020  
18973     0           False          False               2020  
18974     0           False          False               2018  
18975     0           False          False               2020  
18976     0           False          False               2019  
18977     0           False          False               2020  
18978     0           False          False               2019  

[10 rows x 75 columns]